### PR TITLE
Refactoring code to improve reusability

### PR DESCRIPTION
**Pull request by malinthaprasan**
## Purpose
Partially fix: https://github.com/wso2/product-apim-tooling/pull/339

1. Introduced a new package `impl` which resides application logic only.
2. `cmd` package is calling `impl` package for calling the application logic.
3. Moved importAPI/Product/Application, deleteAPI/Product/Application, Listing API/Product/Application logic to impl `while preserving the original commit ownership as much as possible`.
4. Fixed the broken test cases by above changes

### CODE REVIEW

**Comment by npamudika** [comment]
Fixes related to https://github.com/wso2/product-apim-tooling/issues/340

File: import-export-cli/cmd/importAPI.go [hunk 1]
```diff
@@ -82,60 +70,21 @@ var ImportAPICmd = &cobra.Command{
 	Example: importAPICmdExamples,
 	Run: func(cmd *cobra.Command, args []string) {
 		utils.Logln(utils.LogPrefixInfo + importAPICmdLiteral + " called")
-		var apisExportDirectory = filepath.Join(utils.ExportDirectory, utils.ExportedApisDirName)
-
-		cred, err := getCredentials(importEnvironment)
+		credential, err := getCredentials(importEnvironment)
```
Comment (npamudika) [inline import-export-cli/cmd/importAPI.go:73]: Shall we use as credentials?

File: import-export-cli/cmd/importAPI.go [hunk 2]
```diff
@@ -82,60 +70,21 @@ var ImportAPICmd = &cobra.Command{
...
-		return nil, nil, err
-	}
-
-	var buffer []byte
-	if info.IsDir() {
-		_, content, err := resolveYamlOrJson(path.Join(filePath, "Meta-information", "api"))
+		accessOAuthToken, err := credentials.GetOAuthAccessToken(credential, importEnvironment)
 		if err != nil {
-			return nil, nil, err
+			utils.HandleErrorAndExit("Error while getting an access token for importing API", err)
```
Comment (npamudika) [inline import-export-cli/cmd/importAPI.go:79]: Shall we keep an space before printing the err?
```suggestion
			utils.HandleErrorAndExit("Error while getting an access token for importing API ", err)
```

File: import-export-cli/cmd/importAPI.go [hunk 3]
```diff
@@ -82,60 +70,21 @@ var ImportAPICmd = &cobra.Command{
...
-	}
-	api, err := extractAPIDefinition(buffer)
-	if err != nil {
-		return nil, nil, err
-	}
-	return api, buffer, nil
+		err = impl.ImportAPIToEnv(accessOAuthToken, importEnvironment, importAPIFile, importAPIParamsFile, importAPIUpdate,
+			importAPICmdPreserveProvider, importAPISkipCleanup)
+		if err != nil {
+			utils.HandleErrorAndExit("Error importing API", err)
```
Comment (npamudika) [inline import-export-cli/cmd/importAPI.go:84]:
```suggestion
			utils.HandleErrorAndExit("Error importing API ", err)
```

File: import-export-cli/cmd/importAPIProduct.go [hunk 4]
```diff
@@ -75,393 +57,22 @@ var ImportAPIProductCmd = &cobra.Command{
...
-		// We have an HTTP error
-		fmt.Println("Error importing API Product.")
-		fmt.Println("Status: " + resp.Status)
-
-		bodyBuf, err := ioutil.ReadAll(resp.Body)
-		_ = resp.Body.Close()
+		accessOAuthToken, err := credentials.GetOAuthAccessToken(cred, importAPIProductEnvironment)
 		if err != nil {
-			return err
+			utils.HandleErrorAndExit("Error while getting an access token for importing API Product", err)
```
Comment (npamudika) [inline import-export-cli/cmd/importAPIProduct.go:67]:
```suggestion
			utils.HandleErrorAndExit("Error while getting an access token for importing API Product ", err)
```

File: import-export-cli/cmd/importAPIProduct.go [hunk 5]
```diff
@@ -75,393 +57,22 @@ var ImportAPIProductCmd = &cobra.Command{
...
-
-	utils.Logln(utils.LogPrefixInfo + "Creating workspace")
-	tmpPath, err := getTempApiDirectory(resolvedApiProductFilePath)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if importAPIProductSkipCleanup {
-			utils.Logln(utils.LogPrefixInfo+"Leaving", tmpPath)
+			utils.HandleErrorAndExit("Error importing API Product", err)
```
Comment (npamudika) [inline import-export-cli/cmd/importAPIProduct.go:72]:
```suggestion
			utils.HandleErrorAndExit("Error importing API Product ", err)
```

File: import-export-cli/impl/common.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
+package impl
```
Comment (npamudika) [inline import-export-cli/impl/common.go:1]: Missing licence headers?

File: import-export-cli/impl/importApp.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+			fmt.Printf("\nCompleted importing the Application '" + filename + "'\n")
+		} else {
+			fmt.Printf("\nUnable to import the Application\n")
+			fmt.Println("Status: " + resp.Status)
+		}
+
+		//fmt.Println(resp.Header)
+		//resp.Body.Read(bodyContent)
+		//resp.Body.Close()
+		//fmt.Println(bodyContent)
```
Comment (npamudika) [inline import-export-cli/impl/importApp.go:122]: Shall we remove the commented code?

File: import-export-cli/impl/importApp.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+		Transport: tr,
+		Timeout:   time.Duration(utils.HttpRequestTimeout) * time.Second,
+	}
+
+	resp, err := client.Do(req)
+
+	if err != nil {
+		utils.Logln(utils.LogPrefixError, err)
+	} else {
+		//var bodyContent []byte
```
Comment (npamudika) [inline import-export-cli/impl/importApp.go:108]: Shall we remove the commented code?

File: import-export-cli/cmd/importAPIProduct.go (hunk 5 above)
Comment (malinthaprasan) [reply import-export-cli/cmd/importAPIProduct.go:72]: what is the reason for the space..

File: import-export-cli/impl/importApp.go (hunk 7 above)
Comment (malinthaprasan) [reply import-export-cli/impl/importApp.go:122]: sure..

File: import-export-cli/cmd/importAPI.go (hunk 2 above)
Comment (malinthaprasan) [reply import-export-cli/cmd/importAPI.go:79]: why do we need the space here?

Comment (malinthaprasan) [reply import-export-cli/cmd/importAPI.go:79]: Looks like it is already handled in the implementation?

```
func HandleErrorAndContinue(msg string, err error) {
	...
	if err == nil {
		fmt.Fprintf(os.Stderr, "%s: %v\n", ProjectName, msg)
	} else {
		fmt.Fprintf(os.Stderr, "%s: %v Reason: %v\n", ProjectName, msg, err.Error())
	}
}
```

Comment (malinthaprasan) [reply import-export-cli/cmd/importAPI.go:79]: or you mean we need to have 2 spaces here? `..%v<space>Reason..` => `..%v<space><space>Reason..`

File: import-export-cli/cmd/importAPI.go (hunk 1 above)
Comment (malinthaprasan) [reply import-export-cli/cmd/importAPI.go]: Used `cred` as its the variable name used in all the other places.

File: import-export-cli/impl/common.go (hunk 6 above)
Comment (malinthaprasan) [reply import-export-cli/impl/common.go:19]: Fixed

File: import-export-cli/impl/importApp.go (hunk 8 above)
Comment (malinthaprasan) [reply import-export-cli/impl/importApp.go]: Removed

File: import-export-cli/cmd/importAPI.go (hunk 2 above)
Comment (npamudika) [reply import-export-cli/cmd/importAPI.go:79]: No, that should be fine. Didn't notice that we have handled it in the code. Let's keep it as it is.

**Pull request merged**